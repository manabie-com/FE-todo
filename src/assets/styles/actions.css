.Action__btn {
    color: white;
    cursor: pointer;
}
.Action__btn:first-child {
    color: gray;
}
.Action__btn.Action__btn--active {
    background-color: var(--info);
}
.Action__btn.Action__btn--completed {
    background-color: var(--success);
}
.Action__btn.Action__btn--clear {
    background-color: var(--danger);
}
.Action__cbx--check-all {
    margin: 0 12px;
}
