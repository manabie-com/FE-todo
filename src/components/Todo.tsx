import React, { useState } from 'react'
import { Todo } from '../models/todo';
import TodoForm from './TodoForm';


const TodoItem = ({ todo, isTodoCompleted, onUpdateTodoStatus, deleteTodo, updateTodo }:
    { todo: Todo, isTodoCompleted: Function, onUpdateTodoStatus: Function, deleteTodo: Function, updateTodo: Function }
) => {
    const [edit, setEdit] = useState(false);
    const [input, setInput] = useState('');
    const editHandler = () => {
        setEdit(!edit);
    }

    const inputChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
        setInput(e.target.value)
    }

    const deleteHandler = (id: any) => {
        deleteTodo(id)
    }
    const onStatusChangeHandler = (e: React.ChangeEvent<HTMLInputElement>, id: any) => {
        onUpdateTodoStatus(e, id);
    }

    const updateTodoItem = () => {
        updateTodo(todo.id, input);
        setEdit(!edit);
    }
    // if (edit) {
    //     return <TodoForm edit={todo} />
    // }
    return (
        <div className="ToDo__item">
            {/* <input
                type="checkbox"
                checked={isTodoCompleted(todo)}
                onChange={(e) => onStatusChangeHandler(e, todo.id)}
            /> */}
            <span>{edit ? <input
                value={input}
                className="Todo__input"
                placeholder="What need to be done?"
                onChange={inputChangeHandler}
            /> : todo.content}</span>
            <div className="Todo__button">
                <div className="Todo__button edit">
                    {
                        edit ? <button
                            onClick={() => updateTodoItem()}>
                            update
                        </button>
                            :
                            <button
                                onClick={() => editHandler()}>
                                edit
                    </button>
                    }
                </div>
                <div className="Todo__button delete">
                    <button
                        onClick={() => (deleteHandler(todo.id))}
                    >
                        X </button>
                </div>
            </div>
        </div>
    );
}

export default TodoItem
