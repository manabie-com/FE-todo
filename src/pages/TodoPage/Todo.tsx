import React, { useState, useRef } from 'react';
import { useQueryClient } from 'react-query';
import { isTodoCompleted } from 'utils';
import { Todo } from 'models/todo';
import { useUpdateTodo } from 'hooks/todos';
import useOnClickOutside from 'hooks/useClickOutside';

const SingleTodo = ({
  todo,
  onUpdateStatus,
  onDeleteTodo
}: {
  todo: Todo;
  onUpdateStatus: Function;
  onDeleteTodo: Function;
}) => {
  const ref = useRef();

  const queryClient = useQueryClient();
  const [isEdit, setIsEdit] = useState(false);
  const [content, setContent] = useState(() => todo?.content);
  const { updateTodo } = useUpdateTodo();
  useOnClickOutside(ref, () => setIsEdit(false));

  const handleUpdateSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (content) {
      const updatedTodo = { ...todo, content } as Todo;
      updateTodo(updatedTodo, {
        onSuccess: () => {
          queryClient.invalidateQueries('todos');
          setIsEdit(false);
        }
      });
    }
  };

  return (
    <>
      <a className='Todo__item' onDoubleClick={() => setIsEdit(true)} ref={ref}>
        {isEdit ? (
          <form onSubmit={handleUpdateSubmit}>
            <input value={content} onChange={e => setContent(e.target.value)} />
          </form>
        ) : (
          <>
            <input
              type='checkbox'
              checked={isTodoCompleted(todo)}
              onChange={e => onUpdateStatus(e, todo)}
            />
            <span>{todo.content}</span>
            <button
              className='Todo__delete'
              onClick={() => onDeleteTodo(todo.id)}
            >
              X
            </button>
          </>
        )}
      </a>
    </>
  );
};

export default SingleTodo;
