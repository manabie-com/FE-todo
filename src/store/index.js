import { createDataContext } from "../utils";
import Auth from "./logicsAuth";
import Todos from "./logicsTodo";

const initialState = { ...Auth.initialState, ...Todos.initialState };

const reducers = { ...Auth.reducers, ...Todos.reducers };

const actions = { ...Auth.actions, ...Todos.actions };

const reducer = (state, action) => {
  const { type, payload } = action;
  if (typeof reducers[type] === "function") {
    // looking for reducer
    // Call refucer with state, payload, and type
    const newState = reducers[type](state, payload, type);
    if (
      typeof newState === "object" &&
      newState !== null &&
      !Array.isArray(newState)
    ) {
      // return newState
      return newState;
    }
    // return state if reducer is not return a object
    return state;
  }
  //  return state if reducer is not a function
  return state;
};

export const { Context, Provider } = createDataContext(
  reducer,
  actions,
  initialState
);

export default Context;
