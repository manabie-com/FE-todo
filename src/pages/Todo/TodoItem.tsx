import React from 'react';
import { Editable } from '../../components/Editable';
import XIcon from '../../components/icons/XIcon';
import { Todo, TodoStatus } from '../../models/todo';
import { formatDate, isTodoCompleted } from '../../utils';

export const TodoItem: React.FC<
	React.DetailedHTMLProps<
		React.LiHTMLAttributes<HTMLLIElement>,
		HTMLLIElement
	> & {
		todo: Todo;
		onUpdateTodoStatus: (
			e: React.ChangeEvent<HTMLInputElement>,
			todoId: string
		) => void;
		onUpdateTodoContent: (content: string, todoId: string) => void;
		onDeleteTodo: (todoId: string) => void;
	}
> = ({
	todo,
	onUpdateTodoStatus,
	onUpdateTodoContent,
	onDeleteTodo,
	...props
}) => {
	return (
		<li className="Todo__item" {...props}>
			<div>
				<div className="App__checkbox">
					<input
						type="checkbox"
						checked={isTodoCompleted(todo)}
						onChange={(e) => onUpdateTodoStatus(e, todo.id)}
					/>
				</div>
				<div className="Todo__content">
					<Editable
						textStyle={{
							textDecoration:
								todo.status === TodoStatus.COMPLETED ? 'line-through' : 'none',
						}}
						value={todo.content}
						onSave={(content) => onUpdateTodoContent(content, todo.id)}
					/>
					<p>{formatDate(new Date(todo.created_date))}</p>
				</div>
				<button
					type="button"
					className="App__btn__icon App__btn__icon--sm App__btn--danger"
					onClick={() => onDeleteTodo(todo.id)}
				>
					<XIcon />
				</button>
			</div>
		</li>
	);
};
