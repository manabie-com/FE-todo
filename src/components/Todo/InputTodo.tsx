import React, { memo, useContext, useRef } from 'react'
import { useHistory } from 'react-router-dom'

import { TodoContext } from '../../context/TodoContext'
import Service from '../../service'
import { createTodo } from '../../store/actions'

const InputTodo: React.FC = () => {
  const { dispatch } = useContext(TodoContext)
  const inputRef = useRef<HTMLInputElement>(null)
  const history = useHistory()

  const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter' && inputRef.current) {
      try {
        const resp = await Service.createTodo(inputRef.current.value)
        dispatch(createTodo(resp))
        inputRef.current.value = ''
      } catch (e) {
        if (e.response.status === 401) {
          history.push('/')
        }
      }
    }
  }
  return (
    <div className='Todo__creation'>
      <input
        ref={inputRef}
        className='Todo__input'
        placeholder='What need to be done?'
        onKeyDown={onCreateTodo}
      />
    </div>
  )
}

export default memo(InputTodo)
