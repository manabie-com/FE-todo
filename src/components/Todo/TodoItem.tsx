import React, { FC, memo, useContext } from 'react'
import { useHistory } from 'react-router-dom'

import { TodoContext } from '../../context/TodoContext'
import { Todo, TodoStatus } from '../../models/todo'
import Service from '../../service'
import { deleteTodo, updateTodo } from '../../store/actions'
import { isTodoCompleted } from '../../utils'

interface IProps {
  todo: Todo
}

const TodoItem: FC<IProps> = ({ todo }) => {
  const { dispatch } = useContext(TodoContext)
  const history = useHistory()

  const onError = (error: any) => {
    if (error?.response?.status === 401) {
      history.push('/')
    }
  }

  const onToggleComplete = async (e: React.ChangeEvent<HTMLInputElement>) => {
    try {
      const { checked } = e.target
      const nextTodo = await Service.updateTodo({
        ...todo,
        status: checked ? TodoStatus.COMPLETED : TodoStatus.ACTIVE,
      })

      dispatch(updateTodo(nextTodo))
    } catch (error) {
      onError(error)
    }
  }

  const onDelete = async () => {
    try {
      await Service.deleteTodo(todo.id)
      dispatch(deleteTodo(todo.id))
    } catch (error) {
      onError(error)
    }
  }

  return (
    <div className='ToDo__item'>
      <input
        type='checkbox'
        checked={isTodoCompleted(todo)}
        onChange={onToggleComplete}
      />
      <span>{todo.content}</span>
      <button className='Todo__delete' onClick={onDelete}>
        X
      </button>
    </div>
  )
}

export default memo(TodoItem)
