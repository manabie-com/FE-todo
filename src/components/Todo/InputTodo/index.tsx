import React, { memo, useContext, useRef } from 'react'
import { useHistory } from 'react-router-dom'

import { TodoContext } from 'src/context/TodoContext'
import Service from 'src/service/index'
import { createTodo } from 'src/store/actions'
import Input from 'src/components/Input'

import './InputTodo.css'

const InputTodo: React.FC = () => {
  const { dispatch } = useContext(TodoContext)
  const inputRef = useRef<HTMLInputElement>(null)
  const history = useHistory()

  const onCreateTodo = async (e: React.KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter' && inputRef.current) {
      try {
        const resp = await Service.createTodo(inputRef.current.value)
        dispatch(createTodo(resp))
        inputRef.current.value = ''
      } catch (e) {
        if (e.response.status === 401) {
          history.push('/')
        }
      }
    }
  }
  return (
    <div className='Todo__creation'>
      <Input
        ref={inputRef}
        placeholder='What need to be done?'
        onKeyDown={onCreateTodo}
      />
    </div>
  )
}

export default memo(InputTodo)
