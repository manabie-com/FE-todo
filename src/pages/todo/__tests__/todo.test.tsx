import * as React from "react";
import { shallow } from "enzyme";
import Todo, { onDeleteTodoStatus } from "../todo";
import { deleteTodo } from "src/store/actions";

describe("ToDo", () => {
  it("should match a snapshot", () => {
    const routeComponentPropsMock = {
      history: {} as any,
      location: {} as any,
      match: {} as any,
    };
    expect(shallow(<Todo {...routeComponentPropsMock} />)).toMatchSnapshot();
  });
});

describe("onDeleteTodoStatus", () => {
  it("should run correctly", () => {
    const dispatch = jest.fn();
    const id = "id";
    const fn = onDeleteTodoStatus(dispatch);

    fn(id);
    expect(dispatch).toHaveBeenCalledWith(deleteTodo(id));
  });
});
