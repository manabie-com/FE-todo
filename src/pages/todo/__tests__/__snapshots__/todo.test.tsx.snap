// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo should match a snapshot 1`] = `ShallowWrapper {}`;
